File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    total = 0
    ans = 0
    def dfs(self, node: Optional[TreeNode]) -> int:
        if not node: return 0
        s = self.dfs(node.left) + node.val + self.dfs(node.right)
        self.ans = max(self.ans, s * (self.total - s))
        return s
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        self.total = self.dfs(root)
        self.dfs(root)
        return self.ans % 1000000007

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    long ans = 0;
    long total = 0;
    long dfs(TreeNode node) {
        if(node == null) return 0;
        long sum = dfs(node.left) + node.val + dfs(node.right);
        ans = Math.max(ans, sum * (total - sum));
        return sum;
    }
    public int maxProduct(TreeNode root) {
        total = dfs(root);
        dfs(root);
        return (int) (ans % 1000000007);
    }
}